### PR TITLE
feat: implement TOON spec v3.3.0

### LINKED ISSUE #19: TOON Specification Update Available: v3.3.0
## 🔔 TOON Specification Update Detected

A new version of the TOON specification is available from the official repository.

### Version Information

- **Current (Local):** v3.0.3
- **Latest (Upstream):** v3.3.0 (2026-05-21)

### Action Required

Please review the changes in the official specification and update our implementation accordingly:

1. Review the upstream specification: https://github.com/toon-format/spec/blob/main/SPEC.md
2. Update `SPEC.md` to match the latest version
3. Review and update implementation in `nodes/Toon/` if needed
4. Update tests to cover any new features or changes
5. Update `README.md` to reflect the new version
6. Update `package.json` if appropriate

### Diff Summary (first 100 lines)

<details>
<summary>Click to expand diff</summary>

```diff
--- SPEC.md	2026-05-25 10:29:19.386264383 +0000
+++ /tmp/upstream-spec.md	2026-05-25 10:29:20.182259221 +0000
@@ -2,9 +2,9 @@
 
 ## Token-Oriented Object Notation
 
-**Version:** 3.0
+**Version:** 3.3
 
-**Date:** 2025-11-24
+**Date:** 2026-05-21
 
 **Status:** Working Draft
 
@@ -16,11 +16,11 @@
 
 ## Abstract
 
-Token-Oriented Object Notation (TOON) is a line-oriented, indentation-based text format that encodes the JSON data model with explicit structure and minimal quoting. Arrays declare their length and an optional field list once; rows use a single active delimiter (comma, tab, or pipe). Objects use indentation instead of braces; strings are quoted only when required. This specification defines TOON’s concrete syntax, canonical number formatting, delimiter scoping, and strict‑mode validation, and sets conformance requirements for encoders, decoders, and validators. TOON provides a compact, deterministic representation of structured data and is particularly efficient for arrays of uniform objects.
+Token-Oriented Object Notation (TOON) is a line-oriented, indentation-based text format that encodes the JSON data model with explicit structure and minimal quoting. Arrays declare their length and an optional field list once; rows use a single active delimiter (comma, tab, or pipe). Objects use indentation instead of braces; strings are quoted only when required. This specification defines TOON's concrete syntax, canonical number formatting, delimiter scoping, and strict-mode validation, and sets conformance requirements for encoders, decoders, and validators. TOON provides a deterministic representation of structured data, with tabular syntax for arrays of uniform objects.
 
 ## Status of This Document
 
-This document is a Working Draft v3.0 and may be updated, replaced, or obsoleted. Implementers should monitor the canonical repository at https://github.com/toon-format/spec for changes.
+This document is a Working Draft v3.3 and may be updated, replaced, or obsoleted. Implementers should monitor the canonical repository at https://github.com/toon-format/spec for changes.
 
 This specification is stable for implementation but not yet finalized. Breaking changes may occur in future major versions.
 
@@ -57,7 +57,7 @@
 
 ## Table of Contents
 
-- [Introduction](#introduction)
+- [Introduction](#introduction-informative)
 1. [Terminology and Conventions](#1-terminology-and-conventions)
 2. [Data Model](#2-data-model)
 3. [Encoding Normalization (Reference Encoder)](#3-encoding-normalization-reference-encoder)
@@ -74,11 +74,9 @@
 14. [Strict Mode Errors and Diagnostics (Authoritative Checklist)](#14-strict-mode-errors-and-diagnostics-authoritative-checklist)
 15. [Security Considerations](#15-security-considerations)
 16. [Internationalization](#16-internationalization)
-17. [Interoperability and Mappings (Informative)](#17-interoperability-and-mappings-informative)
-18. [IANA Considerations](#18-iana-considerations)
-19. [TOON Core Profile (Normative Subset)](#19-toon-core-profile-normative-subset)
-20. [Versioning and Extensibility](#20-versioning-and-extensibility)
-21. [Intellectual Property Considerations](#21-intellectual-property-considerations)
+17. [IANA Considerations](#17-iana-considerations)
+18. [Versioning and Extensibility](#18-versioning-and-extensibility)
+19. [Intellectual Property Considerations](#19-intellectual-property-considerations)
 
 **Appendices:**
 - [Appendix A: Examples (Informative)](#appendix-a-examples-informative)
@@ -86,16 +84,15 @@
 - [Appendix C: Test Suite and Compliance (Informative)](#appendix-c-test-suite-and-compliance-informative)
 - [Appendix D: Document Changelog (Informative)](#appendix-d-document-changelog-informative)
 - [Appendix E: Acknowledgments and License](#appendix-e-acknowledgments-and-license)
-- [Appendix F: Cross-check With Reference Behavior (Informative)](#appendix-f-cross-check-with-reference-behavior-informative)
-- [Appendix G: Host Type Normalization Examples (Informative)](#appendix-g-host-type-normalization-examples-informative)
+- [Appendix F: Host Type Normalization Examples (Informative)](#appendix-f-host-type-normalization-examples-informative)
 
 ## Introduction (Informative)
 
 ### Purpose and scope
 
-TOON (Token-Oriented Object Notation) is a line-oriented, indentation-based text format that encodes the JSON data model with explicit structure and minimal quoting. It is designed as a compact, deterministic representation of structured data, particularly well-suited to arrays of uniform objects. TOON is often used as a translation layer: produce data as JSON in code, encode to TOON for downstream consumption (e.g., LLM prompts), and decode back to JSON if needed.
+TOON is a token-efficient text format for JSON-shaped data, designed primarily for LLM prompts and contexts where every token costs. It saves tokens vs JSON by declaring array shapes once (length and optional field list) and using indentation in place of braces. TOON is typically used as a translation layer: produce data as JSON in code, encode to TOON for downstream consumption, and decode back to JSON if needed.
 
-### Applicability and non‑goals
+### Applicability and non-goals
 
 Use TOON when:
 - arrays of objects share the same fields (uniform tabular data),
@@ -105,7 +102,7 @@
 
 TOON is not intended to replace:
 - JSON for non-uniform or deeply nested structures where repeated keys are not dominant,
-- CSV for flat, strictly tabular data where maximum compactness is required and nesting is not needed,
+- CSV for flat, strictly tabular data without nesting,
 - general-purpose storage or public APIs. TOON carries the JSON data model; it is a transport/authoring format with explicit structure, not an extended type system or schema language.
 
 Out of scope:
@@ -116,7 +113,7 @@
 ### Relationship to JSON, CSV, and YAML (Informative)
 
 - **JSON**: TOON preserves the JSON data model. It is more compact for uniform arrays of objects by declaring length and fields once. For non-uniform or deeply nested data, JSON may be more efficient.
-- **CSV/TSV**: CSV is typically more compact for flat tables but lacks nesting and type awareness. TOON adds explicit lengths, per-array delimiter scoping, field lists, and deterministic quoting, while remaining lightweight.
+- **CSV/TSV**: CSV is typically more compact for flat tables but lacks nesting and type awareness. TOON adds explicit lengths, per-array delimiter scoping, inline field lists (no separate header row), and deterministic quoting, while remaining lightweight.
 - **YAML**: TOON uses indentation and hyphen markers but is more constrained and deterministic: no comments, explicit array headers with lengths, fixed quoting rules, and a narrow escape set.
 
 ### Example (Informative)
@@ -127,33 +124,18 @@
   2,Bob,user
 ```
 
-### Document roadmap
-
-Normative rules are organized as follows:
-- Data model and canonical number form (§2); normalization on encode (§3); decoding interpretation (§4).
-- Concrete syntax, including root-form determination (§5) and header syntax (§6).
-- Strings and keys (§7); objects (§8); arrays and their sub-forms (§9); objects as list items (§10); delimiter rules (§11).
```

</details>

### Useful Links

- [Official TOON Spec Repository](https://github.com/toon-format/spec)
- [Latest SPEC.md](https://github.com/toon-format/spec/blob/main/SPEC.md)
- [TOON Organization](https://github.com/toon-format)

---

*This issue was automatically created by the [Monitor TOON Specification workflow](.github/workflows/monitor-spec.yml)*

### PR DESCRIPTION
**Pull request by tehw0lf**
## Summary

Implements all normative changes from TOON spec v3.0 → v3.3 (2026-05-21).

- **§7.1 Escaping**: `escapeString` now emits `\uXXXX` for U+0000–U+001F control characters (except `\n`, `\r`, `\t` which keep their named escapes); `unescapeString` now accepts `\uXXXX` (4 hex digits, case-insensitive), rejects lone surrogates (U+D800–U+DFFF), and rejects unknown escape sequences
- **§7.2 Quoting**: `needsQuoting` now covers the full U+0000–U+001F control character range (previously only checked `\n`, `\r`, `\t`)
- **§2 Numbers**: `canonicalizeNumber` now uses exponent notation (lowercase `e`, explicit sign `e+`/`e-`) for numbers with `|n| < 1e-6` or `|n| ≥ 1e21` — the spec changed from "always fixed-point" to "MAY use exponent for out-of-range values"
- **§4/§5 Empty array literal**: bare `[]` at root decodes as empty root array; `key: []` in object field position decodes as an empty array
- **§6 Bracket length**: bracket segments with leading zeros (e.g., `[03]`, `[0001]`) are now a strict-mode error
- **SPEC.md**: replaced with upstream v3.3 verbatim
- `toonSpecVersion` bumped to `3.3.0`

## Test plan

- [ ] All existing tests pass (206 tests, 0 failures)
- [ ] New tests for `\uXXXX` escape (encode + decode)
- [ ] New tests for full U+0000–U+001F `needsQuoting` coverage
- [ ] New tests for exponent notation in `canonicalizeNumber`
- [ ] New tests for `[]` empty array literal decoding
- [ ] New tests for `[03]` leading-zero rejection in strict mode
- [ ] `npm run lint && npm run test && npm run build` all pass

Closes #19

## Summary by CodeRabbit

* **Documentation**
  * Updated TOON specification to v3.3 with revised syntax, delimiter parsing, strict-mode diagnostics, numeric canonicalization, examples, and guidance.

* **New Features**
  * Bare [] is recognized as an empty array literal.
  * Safer safe-mode behavior for key folding and path expansion with stricter segment eligibility.

* **Bug Fixes**
  * Tightened string escape/unescape validation and full U+0000–U+001F control handling.
  * Array-header parsing rejects leading-zero lengths; object/array decoding edge cases clarified.
  * Added strict-mode duplicate-sibling-keys error.

* **Tests**
  * Expanded v3.3 coverage for parsing, escapes, and numeric canonicalization.

* **Chores**
  * Bumped spec metadata to 3.3.0 and updated workflow assignment behavior.

### CODE REVIEW

File: SPEC.md [hunk 1]
```diff
@@ -336,78 +311,89 @@ fieldname     = key
...
   - For object field values (key: value): the document delimiter, even when the object is within an array's scope.
 - It equals "-" or starts with "-" (any hyphen at position 0).
 
 Otherwise, the string MAY be emitted without quotes. Unicode, emoji, and strings with internal (non-leading/trailing) spaces are safe unquoted provided they do not violate the conditions.
 
-### 7.3 Key Encoding (Encoding)
+### 7.3 Key Encoding
 
 Object keys and tabular field names:
 - MAY be unquoted only if they match: ^[A-Za-z_][A-Za-z0-9_.]*$.
```
[the remark this answers is not in the document]
Comment (tehw0lf) [reply SPEC.md:385]: SPEC.md is the upstream document that's kept in the repository as a reference. It's not in scope to make changes to the spec.

File: nodes/Toon/ToonUtils.ts [hunk 2]
```diff
@@ -35,25 +53,47 @@ export function unescapeString(str: string): string {
...
+            throw new Error(`Invalid escape sequence: \\u${hexStr} at position ${i}. Must be 4 hex digits`);
+          }
+          const codePoint = parseInt(hexStr, 16);
+          // Reject lone surrogates (U+D800–U+DFFF)
+          if (codePoint >= 0xd800 && codePoint <= 0xdfff) {
+            throw new Error(`Invalid escape sequence: \\u${hexStr} is a lone surrogate`);
+          }
+          result += String.fromCharCode(codePoint);
+          i += 6;
           break;
```
[the remark this answers is not in the document]
Comment (tehw0lf) [reply nodes/Toon/ToonUtils.ts:90]: For Claude Code: Review this please.
